### PR TITLE
Concurrent db access

### PR DESCRIPTION
**Pull request by kwerber**
This change allows concurrent access to our database again.

Since we currently only use one `Connection` instance for our sqlite database and because the `Connection` instance is not thread safe, allowing parallel transactions can cause problems which is why we currently only allow one thread at a time to access the database.

To allow concurrent access, we need to allow multiple connections to the database,
which in some circumstances, where sqlite does not allow concurrent access (write vs write), 
causes `SQLITE_BUSY` errors.

Waiting until the database is not busy anymore (using the `BUSY_TIMEOUT` option) causes
unnecessary delay, because the waiting thread sometimes only tries again after several hundreds of milliseconds later.
To avoid this delay (and to avoid `SQLITE_BUSY` exceptions in general), a single global write lock is used
to ensure exclusive write access to the entire database.

In addition to this, two new ways to access the database are introduced:

- `DBReadAccess`: Only allows read access and does not acquire any locks
- `DBWriteAccess`: Allows read and write access and does acquire the global lock before any operation is allowed

To manage several database connections, HikariCP is used.

### CODE REVIEW

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/KnownCommitWriteAccess.java [hunk 1]
```diff
@@ -24,17 +24,17 @@ public KnownCommitWriteAccess(DatabaseStorage databaseStorage, TaskWriteAccess t
...
-		databaseStorage.acquireTransaction(db -> {
+		databaseStorage.acquireWriteTransaction(db -> {
 			markCommitsAsKnown(repoId, commits, db);
 			taskAccess.insertTasks(tasks, db);
 		});
 	}
 
 	public void markCommitsAsKnown(RepoId repoId, Collection<CommitHash> commits) {
-		markCommitsAsKnown(repoId, commits, databaseStorage.acquireContext());
+		markCommitsAsKnown(repoId, commits, databaseStorage.acquireWriteAccess());
```
Comment (Garmelon) [inline backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/KnownCommitWriteAccess.java:34]: Why doesn't this use try-with-resources?

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java [hunk 2]
```diff
@@ -66,10 +66,10 @@ public void onTaskDelete(Consumer<TaskId> deleteHandler) {
 	 * @param tasks the tasks to insert
 	 */
 	public void insertTasks(Collection<Task> tasks) {
-		insertTasks(tasks, databaseStorage.acquireContext());
+		insertTasks(tasks, databaseStorage.acquireWriteAccess());
```
Comment (Garmelon) [inline backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java:69]: Why is no try-with-resources used here?

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java [hunk 3]
```diff
@@ -168,10 +166,10 @@ void insertTasks(Collection<Task> originalTasks, DSLContext db) {
 	 * @param taskIds the ids of the tasks to delete
 	 */
 	public void deleteTasks(Collection<TaskId> taskIds) {
-		deleteTasks(taskIds, databaseStorage.acquireContext());
+		deleteTasks(taskIds, databaseStorage.acquireWriteAccess());
```
Comment (Garmelon) [inline backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java:169]: Why is no try-with-resources used here?

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/storage/db/DatabaseStorage.java [hunk 4]
```diff
@@ -76,40 +79,55 @@ private void migrate(String jdbcUrl) {
 	}
 
 	/**
-	 * @return a {@link DSLContext} instance providing jooq functionality along with a connection to
-	 * 	the database
+	 * Acquires read access to the database.
+	 * @return a way to interact with the database in a read only way
```
Comment (Garmelon) [inline backend/backend/src/main/java/de/aaaaaaah/velcom/backend/storage/db/DatabaseStorage.java:83]: A lot of the javadoc in this file seems to not have been auto-formatted.

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/KnownCommitWriteAccess.java (hunk 1 above)
Comment (kwerber) [reply backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/KnownCommitWriteAccess.java]: Fixed in https://github.com/IPDSnelting/velcom/commit/9c85694d224e55f310d9330460dca600ea140c58

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java (hunk 2 above)
Comment (kwerber) [reply backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java]: Fixed in https://github.com/IPDSnelting/velcom/commit/9c85694d224e55f310d9330460dca600ea140c58

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java (hunk 3 above)
Comment (kwerber) [reply backend/backend/src/main/java/de/aaaaaaah/velcom/backend/access/TaskWriteAccess.java]: Fixed in https://github.com/IPDSnelting/velcom/commit/9c85694d224e55f310d9330460dca600ea140c58

File: backend/backend/src/main/java/de/aaaaaaah/velcom/backend/storage/db/DatabaseStorage.java (hunk 4 above)
Comment (kwerber) [reply backend/backend/src/main/java/de/aaaaaaah/velcom/backend/storage/db/DatabaseStorage.java]: Hopefully fixed in https://github.com/IPDSnelting/velcom/commit/6f1f5854816740d8c31bc66b55f552522aa5bafc

**Review by I-Al-Istannen — APPROVED** [review]

**Review by Garmelon — APPROVED** [review]

**Pull request merged**